net-tools/netwatch/src/udp.rs

```diff
@@ -51,97 +68,810 @@ impl UdpSocket {
+        socket_std.connect(addr)?;
+        let socket_tokio = tokio::net::UdpSocket::from_std(socket_std)?;
+        guard.replace((socket_tokio, state));
+        Ok(())
+    }
+
+    /// Returns the local address of this socket.
+    pub fn local_addr(&self) -> std::io::Result<SocketAddr> {
+        let guard = self.socket.read().unwrap();
+        let Some((socket, _)) = guard.as_ref() else {
+            warn!("socket closed");
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::BrokenPipe,
+                "socket closed",
+            ));
+        };
+
+        socket.local_addr()
+    }
+
+    /// Closes the socket, and waits for the underlying `libc::close` call to be finished.
+    pub async fn close(&self) {
+        let socket = self.socket.write().unwrap().take();
+        if let Some((sock, _)) = socket {
+            let std_sock = sock.into_std();
+            let res = tokio::runtime::Handle::current()
+                .spawn_blocking(move || {
+                    // Calls libc::close, which can block
+                    drop(std_sock);
+                })
+                .await;
+            if let Err(err) = res {
+                warn!("failed to close socket: {:?}", err);
+            }
+        }
+    }
+
+    /// Check if this socket is closed.
+    pub fn is_closed(&self) -> bool {
+        self.socket.read().unwrap().is_none()
+    }
+
+    /// Handle potential read errors, updating internal state.
+    ///
+    /// Returns `Some(error)` if the error is fatal otherwise `None.
+    fn handle_read_error(&self, error: std::io::Error) -> Option<std::io::Error> {
+        match error.kind() {
+            std::io::ErrorKind::NotConnected => {
+                // This indicates the underlying socket is broken, and we should attempt to rebind it
+                self.mark_broken();
+                None
+            }
+            _ => Some(error),
+        }
+    }
+
+    /// Handle potential write errors, updating internal state.
+    ///
+    /// Returns `Some(error)` if the error is fatal otherwise `None.
+    fn handle_write_error(&self, error: std::io::Error) -> Option<std::io::Error> {
+        match error.kind() {
+            std::io::ErrorKind::BrokenPipe => {
+                // This indicates the underlying socket is broken, and we should attempt to rebind it
+                self.mark_broken();
+                None
+            }
+            _ => Some(error),
+        }
+    }
+
+    /// Poll for writable
+    pub fn poll_writable(&self, cx: &mut std::task::Context<'_>) -> Poll<std::io::Result<()>> {
+        loop {
+            // check if the socket needs a rebind
+            if self.is_broken() {
+                match self.rebind() {
+                    Ok(()) => {
+                        // all good
+                    }
+                    Err(err) => {
+                        warn!("failed to rebind socket: {:?}", err);
+                        // TODO: improve error
```

Comment:
What still needs to be added to the error?  I think this should be resolved before merging as otherwise it'll be here forever.

Reply:
agreed, I am just not sure what the best way here is